cabal.project.stage3

```diff
@@ -0,0 +1,223 @@
+  xhtml source,
+  Win32 source
+
+
+package *
+  library-vanilla: True
+  shared: False
+  executable-profiling: False
+  executable-dynamic: False
+  executable-static: False
+  -- library-for-ghci will cause a `ld -r` call to create pre-linked objects.
+  -- This helps the internal linker when trying to link (.a) archives with massive
+  -- displacements. In that case the displacement can be in excess of what
+  -- is possible to relocate, and the linker fails. Pre-linking the objects helps with
+  -- this (and linking performance, as we need to process fewer relocations).
+  -- 
+  -- For some cross targets like JavaScript, this will fail as the $LD -r invocation
+  -- might not be properly supported.
+  --
+  --   fs.o: file not recognized: file format not recognized
+  --
+  -- Thus for cross compilers, we outright disable this here. The primary offending
+  -- library is libHSghc, which can easily be 150MB+
+  --
+  -- TODO: this is rather fragile, and should be fixed properly by making pre-linked
+  --             objects of have GHC pre-link excessively large archives on-demand.
+  library-for-ghci: False
+
+if os(wasi)
+  package *
+    shared: True
+  package rts
+    ghc-options: "-optc-DProjectVersion=\"913\""
+    ghc-options: "-optc-DRtsWay=\"v\""
+    ghc-options: "-optc-DHostPlatform=\"wasm32-wasi\""
+    ghc-options: "-optc-DHostArch=\"wasm32\""
+    ghc-options: "-optc-DHostOS=\"unknown\""
+    ghc-options: "-optc-DHostVendor=\"unknown\""
+    ghc-options: "-optc-DBuildPlatform=\"FIXME\""
+    ghc-options: "-optc-DBuildArch=\"FIXME\""
+    ghc-options: "-optc-DBuildOS=\"FIXME\""
+    ghc-options: "-optc-DBuildVendor=\"FIXME\""
+    ghc-options: "-optc-DGhcUnregisterised=\"FIXME\""
+    ghc-options: "-optc-DTablesNextToCode=\"FIXME\""
+    ghc-options: "-optc-DFS_NAMESPACE=rts"
+    ghc-options: "-optl-Wl,--export-dynamic"
+    ghc-options: "-optc-fvisibility=default"
+    ghc-options: "-optc-fvisibility-inlines-hidden"
+    flags: -tables-next-to-code
+
+-- Maybe we should fix this with some import
+if os(linux)
+  package rts
+    ghc-options: "-optc-DProjectVersion=\"913\""
+    ghc-options: "-optc-DRtsWay=\"v\""
+    ghc-options: "-optc-DHostPlatform=\"x86_64-unknown-linux\""
+    ghc-options: "-optc-DHostArch=\"x86_64\""
+    ghc-options: "-optc-DHostOS=\"linux\""
+    ghc-options: "-optc-DHostVendor=\"unknown\""
+    ghc-options: "-optc-DBuildPlatform=\"FIXME\""
+    ghc-options: "-optc-DBuildArch=\"FIXME\""
+    ghc-options: "-optc-DBuildOS=\"FIXME\""
+    ghc-options: "-optc-DBuildVendor=\"FIXME\""
+    ghc-options: "-optc-DGhcUnregisterised=\"FIXME\""
+    ghc-options: "-optc-DTablesNextToCode=\"FIXME\""
+    ghc-options: "-optc-DFS_NAMESPACE=rts"
+    flags: +tables-next-to-code
+
+if os(darwin)
+  package rts
+    ghc-options: "-optc-DProjectVersion=\"913\""
+    ghc-options: "-optc-DRtsWay=\"v\""
+    ghc-options: "-optc-DHostPlatform=\"aarch64-apple-darwin\""
+    ghc-options: "-optc-DHostArch=\"aarch64\""
```

Comment:
If we backport https://gitlab.haskell.org/ghc/ghc/-/merge_requests/14652, there should be less options to pass.